rust-usfm-parser/tests/test_utils.rs

```diff
@@ -0,0 +1,835 @@
+    test_files
+}
+
+fn read_usfm_file(path: &Path) -> Result<String, std::io::Error> {
+    fs::read_to_string(path)
+}
+
+pub fn is_valid_usfm<P: AsRef<Path>>(
+    input_usfm_path: P,
+) -> Result<bool, Box<dyn std::error::Error>> {
+    // Check override list first
+    let path_str = input_usfm_path
+        .as_ref()
+        .to_str()
+        .ok_or("Invalid path encoding")?;
+
+    if let Some(&override_value) = PASS_FAIL_OVERRIDE_LIST.get(path_str) {
+        return Ok(override_value);
+    }
+
+    // Get metadata file path
+    let meta_file_path = path_str.replace("origin.usfm", "metadata.xml");
+    let mut meta_xml_string = fs::read_to_string(&meta_file_path)?;
+
+    // Remove XML declaration if present
+    if meta_xml_string.starts_with("<?xml ") {
+        if let Some(pos) = meta_xml_string.find('\n') {
+            meta_xml_string = meta_xml_string[pos + 1..].to_string();
+        }
+    }
+
+    // Simple XML parsing just to find the validated tag's content
+    let doc = roxmltree::Document::parse(&meta_xml_string)?;
+    let root = doc.root_element();
+
+    // Find the validated node and check its text
+    if let Some(validated_node) = root.children().find(|n| n.has_tag_name("validated")) {
+        if let Some(text) = validated_node.text() {
+            return Ok(text != "fail");
+        }
+    }
+
+    // If we didn't find a validated tag or it had no text, default to true
+    Ok(true)
+}
+
+
+pub fn get_negative_tests() -> Result<Vec<PathBuf>, Box<dyn std::error::Error>> {
+    let mut negative_tests = Vec::new();
+
+    // Get all USFM files using the test patterns
+    let test_files = get_test_files();
+
+    // Check each file and collect invalid ones
+    for file_path in test_files {
+        if !is_valid_usfm(&file_path)? {
+            negative_tests.push(file_path);
+        }
+    }
+
+    Ok(negative_tests)
+}
+
+pub const EXCLUDE_USX_FILES: &[&str] = &[
+    // \ef not treated as inline content of paragraph
+    "tests/specExamples/extended/contentCatogories2/origin.xml",
+    // verse number="+"!!!
+    "tests/specExamples/extended/sectionIntroductions/origin.xml",
+    // lit element treated as a body paragraph enclosing a verse!
+    "tests/specExamples/character/origin.xml",
+    // last verse text given outside of paragraph
+    "tests/usfmjsTests/esb/origin.xml",
+    // ~ not being replaced by nbsp in usfm-grammar
+    "tests/special-cases/nbsp/origin.xml",
+    // attributes treated as text content of marker
+    "tests/special-cases/empty-attributes/origin.xml",
+    "tests/biblica/CategoriesOnNotes/origin.xml",
+    "tests/biblica/CrossRefWithPipe/origin.xml",
+    // ref node has type ref. Is it char or ref?
+    "tests/usfmjsTests/usfmBodyTestD/origin.xml",
+    // \v and other contents contained inside \lit. New docs doesnt have \lit
+    "tests/usfmjsTests/usfm-body-testF/origin.xml",
+];
+
+// Test functions
+#[cfg(test)]
+mod tests {
+    use rust_usfm::usj_generator::usj_generator;
```

Comment:
it would be good to keep tests in a separate file and keep only utility functions here. As we add more testing code it we help us maintain modularity.

Reply:
test_usj.rs would be good.